pulp_docker/app/views.py

```diff
@@ -14,3 +57,235 @@ def get(self, request):
+        """
+        Dispatch to the appropriate responder (method).
+
+        Args:
+            path (str): The fully qualified path to the file to be served.
+
+        Returns:
+            django.http.StreamingHttpResponse: on found.
+            django.http.HttpResponseNotFound: on not-found.
+            django.http.HttpResponseForbidden: on forbidden.
+            django.http.HttpResponseRedirect: on redirect to the streamer.
+        """
+        server = settings.CONTENT['WEB_SERVER']
+
+        try:
+            responder = self.RESPONDER[server]
+        except KeyError:
+            raise ValueError(_('Web server "{t}" not supported.').format(t=server))
+        else:
+            return responder(self, path, headers)
+
+    # Mapping of responder-method based on the type of web server.
+    RESPONDER = {
+        'django': _django,
+        'apache': _apache,
+        'nginx': _nginx,
+    }
+
+
+class TagView(views.APIView, ServeContentMixin):
+    """
+    APIView for Docker Registry v2 tag API
+    """
+
+    authentication_classes = []
+    permission_classes = []
+    content_negotiation_class = IgnoreClientContentNegotiation
+
+    def get(self, request, path, tag_name):
+        """
+        Return a response to the "GET" action.
+        """
+        distribution = get_object_or_404(DockerDistribution, base_path=path)
+        try:
+            ca = ContentArtifact.objects.get(
+                content__in=distribution.publication.repository_version.content,
+                relative_path=tag_name)
+            content = ca.content.cast()
+            if content.manifest:
+                headers = {'Content-Type': content.manifest.media_type}
+            else:
+                headers = {'Content-Type': content.manifest_list.media_type}
+
+        except MultipleObjectsReturned:
+            log.debug(
+                _('Multiple (pass-through) matches for {b}/{p}'),
+                {
+                    'b': distribution.base_path,
+                    'p': tag_name,
+                }
+            )
+
+        except ObjectDoesNotExist:
+            pass
+        else:
+            artifact = ca.artifact
+            if artifact:
+                return self._dispatch(artifact.file.name, headers)
+            else:
+                raise ArtifactNotFound(path)
+
+        raise PathNotResolved(path)
+
+
+class BlobManifestView(views.APIView, ServeContentMixin):
+    """
+    APIView for Docker Registry v2 blob and manifest APIs
+    """
+
+    authentication_classes = []
+    permission_classes = []
+    content_negotiation_class = IgnoreClientContentNegotiation
+
+    def get(self, request, path, digest):
+        """
+        Return a response to the "GET" action.
+        """
+        distribution = get_object_or_404(DockerDistribution, base_path=path)
+        try:
+            ca = ContentArtifact.objects.get(
+                content__in=distribution.publication.repository_version.content,
```

Comment:
`content__in` might be a rough db query. If we this is slow, maybe it would be better to do 

    ContentArtifact.objects.filter(relative_path=digest).filter(content__in=...)`

Django might be smart enough to do optimize this under the hood though.